File: tests/initramfs/initramfs_test.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct expected_entry {
  const char *name;
  uint32_t mode;
  uint32_t nlink;
  const char *body;
  uint32_t rdev_major;
  uint32_t rdev_minor;
};

static const struct expected_entry expected[] = {
    {"bin", 0040755, 2, "", 0, 0},
    {"dev", 0040755, 2, "", 0, 0},
    {"empty", 0040755, 2, "", 0, 0},
    {"empty/dir", 0040755, 2, "", 0, 0},
    {"etc", 0040755, 2, "", 0, 0},
    {"usr", 0040755, 2, "", 0, 0},
    {"usr/bin", 0040755, 2, "", 0, 0},
    {"etc/config", 0100644, 1, "config\n", 0, 0},
    {"bin/tool", 0100755, 1, "tool\n", 0, 0},
    {"init", 0100755, 1, "tool\n", 0, 0},
    {"usr/bin/tool", 0100755, 1, "tool\n", 0, 0},
    {"bin/config", 0120777, 1, "/etc/config", 0, 0},
    {"dev/null", 0020666, 1, "", 1, 3},
    {"TRAILER!!!", 0, 1, "", 0, 0},
};

static int fail(const char *message, const char *name) {
  fprintf(stderr, "initramfs_test: %s: %s\n", message, name);
  return 0;
}

static int read_exact(FILE *file, void *data, size_t size) {
  return fread(data, 1, size, file) == size;
}

static int read_hex(const char *text, uint32_t *value) {
  char field[9];
  char *end;
  unsigned long parsed;
  memcpy(field, text, 8);
  field[8] = '\0';
  parsed = strtoul(field, &end, 16);
  if (*end != '\0' || parsed > UINT32_MAX)
    return 0;
  *value = (uint32_t)parsed;
  return 1;
}

static int skip_padding(FILE *file, uint64_t *offset, uint32_t alignment) {
  unsigned char padding[512];
  size_t size = (size_t)((alignment - *offset % alignment) % alignment);
  size_t i;
  if (!read_exact(file, padding, size))
    return 0;
  for (i = 0; i < size; ++i) {
    if (padding[i] != 0)
      return 0;
  }
  *offset += size;
  return 1;
}

static int check_entry(FILE *file, uint64_t *offset, size_t index) {
  const struct expected_entry *want = &expected[index];
  char header[110];
  uint32_t fields[13];
  uint32_t name_size;
  uint32_t body_size;
  char *name;
  char *body;
  size_t i;

  if (!read_exact(file, header, sizeof(header)) ||
      memcmp(header, "070701", 6) != 0)
    return fail("invalid newc header", want->name);
  *offset += sizeof(header);
  for (i = 0; i < 13; ++i) {
    if (!read_hex(header + 6 + i * 8, &fields[i]))
      return fail("invalid newc field", want->name);
  }
  name_size = fields[11];
  body_size = fields[6];
  name = malloc(name_size ? name_size : 1);
  body = malloc(body_size ? body_size : 1);
  if (!name || !body)
    return fail("out of memory", want->name);
  if (!name_size || !read_exact(file, name, name_size) ||
      name[name_size - 1] != '\0')
    return fail("invalid entry name", want->name);
  *offset += name_size;
  if (!skip_padding(file, offset, 4) ||
      (body_size && !read_exact(file, body, body_size)))
    return fail("truncated entry", want->name);
  *offset += body_size;
  if (!skip_padding(file, offset, 4))
    return fail("invalid entry padding", want->name);

  if (name_size != strlen(want->name) + 1 || strcmp(name, want->name) != 0 ||
      fields[0] != (uint32_t)index + 1 || fields[1] != want->mode ||
      fields[2] != 0 || fields[3] != 0 || fields[4] != want->nlink ||
      fields[5] != 0 || fields[7] != 0 || fields[8] != 0 ||
      fields[9] != want->rdev_major || fields[10] != want->rdev_minor ||
      fields[12] != 0 || body_size != strlen(want->body) ||
      memcmp(body, want->body, body_size) != 0) {
    free(name);
    free(body);
    return fail("entry does not match", want->name);
  }
  free(name);
  free(body);
  return 1;
}

static int compare_archives(const char *first_path, const char *second_path) {
  FILE *first = fopen(first_path, "rb");
  FILE *second = fopen(second_path, "rb");
  int first_byte;
  int second_byte;
  if (!first || !second) {
    if (first)
      fclose(first);
    if (second)
      fclose(second);
    return fail("could not open archive for comparison", second_path);
  }
  do {
    first_byte = fgetc(first);
    second_byte = fgetc(second);
  } while (first_byte != EOF && first_byte == second_byte);
  if (ferror(first) || ferror(second) || first_byte != second_byte) {
    fclose(first);
    fclose(second);
    return fail("reordered inputs produced different archives", second_path);
  }
  if (fclose(first) != 0 || fclose(second) != 0)
    return fail("could not close compared archives", second_path);
  return 1;
}

int main(int argc, char **argv) {
  FILE *file;
  uint64_t offset = 0;
  size_t i;
  int byte;
  if (argc != 3) {
    fail("expected two archive paths", "argv");
    return 1;
  }
  file = fopen(argv[1], "rb");
  if (!file) {
    fail("could not open archive", argv[1]);
    return 1;
  }
  for (i = 0; i < sizeof(expected) / sizeof(expected[0]); ++i) {
    if (!check_entry(file, &offset, i)) {
      fclose(file);
      return 1;
    }
  }
  while ((byte = fgetc(file)) != EOF) {
    if (byte != 0) {
      fclose(file);
      fail("nonzero final padding", argv[1]);
      return 1;
    }
    ++offset;
  }
  if (ferror(file) || offset % 512 != 0) {
    fclose(file);
    fail("invalid archive size", argv[1]);
    return 1;
  }
  if (fclose(file) != 0) {
    fail("could not close archive", argv[1]);
    return 1;
  }
  return compare_archives(argv[1], argv[2]) ? 0 : 1;
}

File: tests/initramfs/initramfs_test.go
package initramfs_test

import (
	"bytes"
	"flag"
	"fmt"
	"os"
	"strconv"
	"testing"

	"github.com/bazelbuild/rules_go/go/runfiles"
)

const (
	newcHeaderSize = 110
	newcMagic      = "070701"
)

type newcEntry struct {
	name      string
	inode     uint32
	mode      uint32
	uid       uint32
	gid       uint32
	nlink     uint32
	mtime     uint32
	devMajor  uint32
	devMinor  uint32
	rdevMajor uint32
	rdevMinor uint32
	check     uint32
	body      []byte
}

var expectedEntries = []newcEntry{
	{name: "bin", inode: 1, mode: 0o040755, nlink: 2},
	{name: "dev", inode: 2, mode: 0o040755, nlink: 2},
	{name: "empty", inode: 3, mode: 0o040755, nlink: 2},
	{name: "empty/dir", inode: 4, mode: 0o040755, nlink: 2},
	{name: "etc", inode: 5, mode: 0o040755, nlink: 2},
	{name: "usr", inode: 6, mode: 0o040755, nlink: 2},
	{name: "usr/bin", inode: 7, mode: 0o040755, nlink: 2},
	{name: "etc/config", inode: 8, mode: 0o100644, nlink: 1, body: []byte("config\n")},
	{name: "bin/tool", inode: 9, mode: 0o100755, nlink: 1, body: []byte("tool\n")},
	{name: "init", inode: 10, mode: 0o100755, nlink: 1, body: []byte("tool\n")},
	{name: "usr/bin/tool", inode: 11, mode: 0o100755, nlink: 1, body: []byte("tool\n")},
	{name: "bin/config", inode: 12, mode: 0o120777, nlink: 1, body: []byte("/etc/config")},
	{name: "dev/null", inode: 13, mode: 0o020666, nlink: 1, rdevMajor: 1, rdevMinor: 3},
	{name: "TRAILER!!!", inode: 14, nlink: 1},
}

func TestInitramfs(t *testing.T) {
	paths := flag.Args()
	if len(paths) == 0 {
		t.Skip("archive inputs are supplied by the Bazel target")
	}
	if len(paths) != 2 {
		t.Fatalf("got %d archive arguments, want 2", len(paths))
	}

	archive := readArchive(t, paths[0])
	reordered := readArchive(t, paths[1])
	if !bytes.Equal(archive, reordered) {
		t.Fatal("reordered inputs produced a different archive")
	}
	if len(archive)%512 != 0 {
		t.Fatalf("archive size %d is not aligned to 512 bytes", len(archive))
	}

	entries, offset, err := parseNewc(archive)
	if err != nil {
		t.Fatal(err)
	}
	if len(entries) != len(expectedEntries) {
		t.Fatalf("got %d entries, want %d", len(entries), len(expectedEntries))
	}
	for i := range expectedEntries {
		if err := compareEntry(entries[i], expectedEntries[i]); err != nil {
			t.Errorf("entry %d: %v", i, err)
		}
	}
	if padding := archive[offset:]; !allZero(padding) {
		t.Fatalf("final %d-byte padding contains nonzero data", len(padding))
	}
}

func readArchive(t *testing.T, path string) []byte {
	t.Helper()
	resolved, err := runfiles.Rlocation(path)
	if err != nil {
		t.Fatalf("resolve %q: %v", path, err)
	}
	data, err := os.ReadFile(resolved)
	if err != nil {
		t.Fatalf("read %q: %v", path, err)
	}
	return data
}

func parseNewc(archive []byte) ([]newcEntry, int, error) {
	var entries []newcEntry
	offset := 0
	for {
		if len(archive)-offset < newcHeaderSize {
			return nil, 0, fmt.Errorf("truncated header at offset %d", offset)
		}
		header := archive[offset : offset+newcHeaderSize]
		if string(header[:6]) != newcMagic {
			return nil, 0, fmt.Errorf("invalid magic %q at offset %d", header[:6], offset)
		}
		offset += newcHeaderSize

		fields := make([]uint32, 13)
		for i := range fields {
			value, err := strconv.ParseUint(string(header[6+i*8:14+i*8]), 16, 32)
			if err != nil {
				return nil, 0, fmt.Errorf("field %d at offset %d: %w", i, offset-newcHeaderSize, err)
			}
			fields[i] = uint32(value)
		}

		nameSize := int(fields[11])
		if nameSize < 1 || nameSize > len(archive)-offset {
			return nil, 0, fmt.Errorf("invalid name size %d at offset %d", nameSize, offset)
		}
		nameBytes := archive[offset : offset+nameSize]
		if nameBytes[nameSize-1] != 0 || bytes.IndexByte(nameBytes[:nameSize-1], 0) >= 0 {
			return nil, 0, fmt.Errorf("invalid NUL-terminated name at offset %d", offset)
		}
		offset += nameSize
		var err error
		offset, err = skipZeroPadding(archive, offset, 4)
		if err != nil {
			return nil, 0, err
		}

		bodySize := int(fields[6])
		if bodySize > len(archive)-offset {
			return nil, 0, fmt.Errorf("truncated body at offset %d: need %d bytes", offset, bodySize)
		}
		body := bytes.Clone(archive[offset : offset+bodySize])
		offset += bodySize
		offset, err = skipZeroPadding(archive, offset, 4)
		if err != nil {
			return nil, 0, err
		}

		entry := newcEntry{
			name:      string(nameBytes[:nameSize-1]),
			inode:     fields[0],
			mode:      fields[1],
			uid:       fields[2],
			gid:       fields[3],
			nlink:     fields[4],
			mtime:     fields[5],
			devMajor:  fields[7],
			devMinor:  fields[8],
			rdevMajor: fields[9],
			rdevMinor: fields[10],
			check:     fields[12],
			body:      body,
		}
		entries = append(entries, entry)
		if entry.name == "TRAILER!!!" {
			return entries, offset, nil
		}
	}
}

func skipZeroPadding(data []byte, offset, alignment int) (int, error) {
	padding := (alignment - offset%alignment) % alignment
	if padding > len(data)-offset {
		return 0, fmt.Errorf("truncated padding at offset %d", offset)
	}
	if !allZero(data[offset : offset+padding]) {
		return 0, fmt.Errorf("nonzero padding at offset %d", offset)
	}
	return offset + padding, nil
}

func allZero(data []byte) bool {
	for _, value := range data {
		if value != 0 {
			return false
		}
	}
	return true
}

func compareEntry(got, want newcEntry) error {
	if got.name != want.name {
		return fmt.Errorf("name = %q, want %q", got.name, want.name)
	}
	if got.inode != want.inode {
		return fmt.Errorf("%s: inode = %d, want %d", got.name, got.inode, want.inode)
	}
	if got.mode != want.mode || got.uid != want.uid || got.gid != want.gid ||
		got.nlink != want.nlink || got.mtime != want.mtime ||
		got.devMajor != want.devMajor || got.devMinor != want.devMinor ||
		got.rdevMajor != want.rdevMajor || got.rdevMinor != want.rdevMinor ||
		got.check != want.check {
		return fmt.Errorf(
			"%s: metadata = (mode=%#o uid=%d gid=%d nlink=%d mtime=%d dev=%d:%d rdev=%d:%d check=%d), want (%#o,%d,%d,%d,%d,%d:%d,%d:%d,%d)",
			got.name,
			got.mode, got.uid, got.gid, got.nlink, got.mtime, got.devMajor, got.devMinor, got.rdevMajor, got.rdevMinor, got.check,
			want.mode, want.uid, want.gid, want.nlink, want.mtime, want.devMajor, want.devMinor, want.rdevMajor, want.rdevMinor, want.check,
		)
	}
	if !bytes.Equal(got.body, want.body) {
		return fmt.Errorf("%s: body = %q, want %q", got.name, got.body, want.body)
	}
	return nil
}
